src/dune_engine/build_system.ml

```diff
@@ -127,20 +132,20 @@ module Alias0 = struct
   open Action_builder.O
 
   let dep_rec_internal ~name ~dir ~ctx_dir =
-    let f dir acc =
+    let module M =
+      File_tree.Dir.Make_map_reduce (Action_builder) (Monoid.Exists)
```

Comment:
Shall we move this somewhere to top-level? I'm always wondering if the compiler is clever enough to float such definitions automatically, and I guess we don't want to allocate a new record on each call of this function.

Reply:
Btw, `Make_map_reduce (Action_builder) (Monoid.Exists)` looks pretty neat!

Reply:
Indeed, something lit up in my brain when I came up with this :) I'm loving monads and monoids now.

Reply:
Regarding perfs, seems worth looking at the cmm to see what the compiler is doing.

Reply:
We had a look, and indeed the functor is applied on every call to `dep_rec_internal`, which doesn't feel very efficient. Changing this code.